torch/distributed/device_mesh.py

```diff
@@ -216,11 +222,25 @@ def __init__(
                 if isinstance(mesh, torch.Tensor)
                 else torch.tensor(mesh, dtype=torch.int)
             )
-            self.mesh_dim_names = mesh_dim_names
+            self.mesh_dim_names = tuple(mesh_dim_names) if mesh_dim_names else None
 
             # private field to pre-generate DeviceMesh's hash
             self._flatten_mesh_list = tuple(self.mesh.flatten().tolist())
-            self._hash = hash((self._flatten_mesh_list, self.mesh.shape, id(self)))
+            hash_obj = (
+                (
+                    self._flatten_mesh_list,
+                    self.mesh.shape,
+                    self.device_type,
+                    self.mesh_dim_names,
+                )
+                if self.mesh_dim_names
+                else (
+                    self._flatten_mesh_list,
+                    self.mesh.shape,
+                    self.device_type,
+                )
```

Comment:
Wonder if you actually need this else? Seems like you could just put `mesh_dim_names` into the obj?